### PR TITLE
HDDS-1511. Space tracking for Open Containers in HDDS Volumes. Contributed by Supratim Deka

### PR DESCRIPTION
**Pull request by supratimdeka**
Fixed junit failures. Addressed comment from Arpit Agarwal.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 380 | trunk passed |
| +1 | compile | 190 | trunk passed |
| +1 | checkstyle | 50 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 788 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 115 | trunk passed |
| 0 | spotbugs | 236 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 412 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 26 | Maven dependency ordering for patch |
| +1 | mvninstall | 386 | the patch passed |
| +1 | compile | 212 | the patch passed |
| +1 | javac | 212 | the patch passed |
| +1 | checkstyle | 59 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 676 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 120 | the patch passed |
| +1 | findbugs | 441 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 127 | hadoop-hdds in the patch failed. |
| -1 | unit | 1382 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 6287 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestHybridPipelineOnDatanode |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineProvider |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-812/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/812 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 99eda205eb5d 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b832e17 |
| Default Java | 1.8.0_191 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/1/testReport/ |
| Max. process+thread count | 4448 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java [hunk 1]
```diff
@@ -280,6 +298,37 @@ public synchronized void closeContainer() {
...
+    long unused = getMaxSize() - getBytesUsed();
+
+    // only if container size < max size
+    if (unused > 0 && committedSpace) {
+      getVolume().incCommittedBytes(0 - unused);
+    }
+    committedSpace = false;
+  }
+
+  public void commitSpace() {
```
Comment (jiwq) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java]: Public method need a JavaDoc comment

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java [hunk 2]
```diff
@@ -321,10 +370,20 @@ public long getWriteBytes() {
 
   /**
    * Increase the number of bytes write into the container.
+   * Also decrement committed bytes against the bytes written.
    * @param bytes the number of bytes write into the container.
    */
   public void incrWriteBytes(long bytes) {
+    long unused = getMaxSize() - getBytesUsed();
```
Comment (jiwq) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java]: Maybe we need a new method to decrement the bytes  such as named `decrWriteBytes`.

Comment (supratimdeka) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java]: there is no requirement for decrWriteBytes currently in any flow.
But I have added a new method decCommittedBytes(delta) in HddsVolume to complement incCommittedBytes(delta). This improves readability and the caller does not have to do a (0 - delta) to decrease commit space.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java (hunk 1 above)
Comment (supratimdeka) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java]: will add in the next commit

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java (hunk 2 above)
Comment (jiwq) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java]: Yes, it's good!

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 27 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 60 | Maven dependency ordering for branch |
| +1 | mvninstall | 392 | trunk passed |
| +1 | compile | 198 | trunk passed |
| +1 | checkstyle | 48 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 773 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 128 | trunk passed |
| 0 | spotbugs | 245 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 422 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for patch |
| +1 | mvninstall | 452 | the patch passed |
| +1 | compile | 192 | the patch passed |
| +1 | javac | 192 | the patch passed |
| +1 | checkstyle | 48 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 618 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 121 | the patch passed |
| +1 | findbugs | 432 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 132 | hadoop-hdds in the patch failed. |
| -1 | unit | 1363 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 40 | The patch does not generate ASF License warnings. |
| | | 5585 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.client.rpc.TestCommitWatcher |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-812/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/812 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux efc050fcffb4 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 2d8282b |
| Default Java | 1.8.0_191 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/2/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/2/testReport/ |
| Max. process+thread count | 4897 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by arp7** [comment]
+1. Added _ozone_ label to trigger full precommit run.

**Comment by arp7** [comment]
@supratimdeka the jenkins run flagged 3 unit test failures. Can you please confirm they are unrelated to the patch by running them locally?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 29 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 61 | Maven dependency ordering for branch |
| +1 | mvninstall | 403 | trunk passed |
| +1 | compile | 195 | trunk passed |
| +1 | checkstyle | 44 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 741 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 118 | trunk passed |
| 0 | spotbugs | 238 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 418 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| +1 | mvninstall | 395 | the patch passed |
| +1 | compile | 198 | the patch passed |
| +1 | javac | 198 | the patch passed |
| +1 | checkstyle | 51 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 620 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 118 | the patch passed |
| +1 | findbugs | 453 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 151 | hadoop-hdds in the patch failed. |
| -1 | unit | 1167 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 5346 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.hdds.scm.pipeline.TestSCMPipelineManager |
|   | hadoop.ozone.om.TestOzoneManagerHA |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-812/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/812 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux d8645cd88eb6 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 2d8282b |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/3/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/3/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/3/testReport/ |
| Max. process+thread count | 4679 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-812/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by supratimdeka** [comment]
TestBlockOutputStreamWithFailures appears to be unpredictable - failures are intermittent.

TestHddsDatanodeService appears to be broken.
both are failing on trunk without the patch. Filed a jira for this issue.

remaining tests are passing locally.

**Pull request merged**